File: solutions/ruby/hamming/1/hamming.rb
class StrandLengthError < ArgumentError
  def initialize(message = 'Strand lengths must equal')
    super
  end
end

class Hamming

  def self.compute(strand1, strand2)
    new(strand1, strand2).distance
  end

  private

  attr_accessor :nucleotides

  def initialize(strand1, strand2)
    raise StrandLengthError if strand1.size != strand2.size
    self.nucleotides = strand1.chars.zip(strand2.chars)
  end

  public

  def distance
    nucleotides.count { |n1, n2| n1 != n2 }
  end

end

File: solutions/elixir/hamming/1/lib/hamming.ex
defmodule Hamming do

  def hamming_distance(strand1, strand2)
    when length(strand1) != length(strand2) do
      {:error, "strands must be of equal length"}
  end

  def hamming_distance(strand1, strand2) do
    count = Enum.zip(strand1, strand2)
    |> Enum.count(fn {nuc1, nuc2} -> nuc1 != nuc2 end)

    {:ok, count}
  end

end
